src/pybind/rados/rados.pyx

```diff
@@ -3525,6 +3526,50 @@ returned %d, but should return zero on success." % (self.name, ret))
             free(_values)
             free(_lens)
 
+    @requires(('write_op', WriteOp), ('keys', tuple), ('values', tuple), ('key_val_num', int))
+    def set_omap2(self, write_op, keys, values, key_val_num):
```

Comment:
can you explain the functionality and use case for this function?

Reply:
> can you explain the functionality and use case for this function?

@zjcmszh Thanks!

Reply:
set_omap2 can use key_val_num to specify the number of key-value pairs

Reply:
```
**
 * Set key/value pairs on an object
 *
 * @param write_op operation to add this action to
 * @param keys array of null-terminated char arrays representing keys to set
 * @param vals array of pointers to values to set
 * @param lens array of lengths corresponding to each value
 * @param num number of key/value pairs to set
 */
CEPH_RADOS_API void rados_write_op_omap_set(rados_write_op_t write_op,
                                            char const* const* keys,
                                            char const* const* vals,
                                            const size_t *lens,
                                            size_t num);


/**
 * Set key/value pairs on an object
 *
 * @param write_op operation to add this action to
 * @param keys array of null-terminated char arrays representing keys to set
 * @param vals array of pointers to values to set
 * @param key_lens array of lengths corresponding to each key
 * @param val_lens array of lengths corresponding to each value
 * @param num number of key/value pairs to set
 */
CEPH_RADOS_API void rados_write_op_omap_set2(rados_write_op_t write_op,
                                            char const* const* keys,
                                            char const* const* vals,
                                            const size_t *key_lens,
                                            const size_t *val_lens,
                                            size_t num);
```
set_omap()  must have the same number of key and val.
 keys = ("1", "2", "3", "4", "5")
 values = (b"a", b"bb", b"ccc", b"dddd")

set_omap2  can have different number of key and val.
 keys = ("1", "2", "3", "4", "5")
 values = (b"a", b"bb",)

Reply:
```
   def test_set_omap2(self):
        keys = ("1", "2", "3", "4", "5")
        values = (b"a", b"bb")
        with WriteOpCtx() as write_op:
            self.ioctx.set_omap2(write_op, keys, values, 2)
            self.ioctx.operate_write_op(write_op, "test_obj")

    def test_set_omap(self):
        keys = ("1", "2", "3", "4")
        values = (b"a", b"bb", b"ccc", b"dddd")
        with WriteOpCtx() as write_op:
            self.ioctx.set_omap(write_op, keys, values)
            self.ioctx.operate_write_op(write_op, "test_obj1")
```

Reply:
Thanks, but we don’t think so set_omap2 is meant for a different number of key/val pairs. 
You can see the commit 253dfd5f8f8b80838127c7326a5f821c424d2894 that specifies: 

> OMAP accept key values that  are NUL-terminated. This makes them unsuitable for operation on keys that have NULs embedded in them (like those produced by librbd).

But python2 does not support any byte type object. So we don’t see the functionality of this function in python binding. That’s our concern.

Reply:
What do you think is the purpose of rados_write_op_omap_set2()'s existence?

Reply:
from my understanding, this function was intended for those key/val pairs which have null values the set_omap was using null character to determine the end of keys/values so if an element in keys/vals is a null we won't be able to use such cases, which would be more relevant for datatypes such as binary, but python2 does converts the binary data to string, which kills the purpose of this function, if you think otherwise, please explain.